File: bms/go/geo/geo_s1_Interval_Intersects/geo_s1_Interval_Intersects.go
package test

import (
	"math"
)

func (i geo_s1_Interval) Intersects(oi geo_s1_Interval) bool {
	if i.IsEmpty() || oi.IsEmpty() {
		return false
	}
	if i.IsInverted() {
		return oi.IsInverted() || oi.Lo <= i.Hi || oi.Hi >= i.Lo
	}
	if oi.IsInverted() {
		return oi.Lo <= i.Hi || oi.Hi >= i.Lo
	}
	return oi.Lo <= i.Hi && oi.Hi >= i.Lo
}

func (i geo_s1_Interval) IsEmpty() bool { return i.Lo == math.Pi && i.Hi == -math.Pi }

func (i geo_s1_Interval) IsInverted() bool { return i.Lo > i.Hi }

type geo_s1_Interval struct {
	Lo, Hi float64
}

File: translations_only/success/go/geo/geo_s1_Interval_Intersects/mistral/CAPR/geo_s1_Interval_Intersects.rs

use std::f64;

struct GeoS1Interval {
    lo: f64,
    hi: f64,
}

impl GeoS1Interval {
    fn is_empty(&self) -> bool {
        self.lo == f64::consts::PI && self.hi == -f64::consts::PI
    }

    fn is_inverted(&self) -> bool {
        self.lo > self.hi
    }
}

fn intersects(i: &GeoS1Interval, oi: &GeoS1Interval) -> bool {
    if i.is_empty() || oi.is_empty() {
        return false;
    }
    if i.is_inverted() {
        return oi.is_inverted() || oi.lo <= i.hi || oi.hi >= i.lo;
    }
    if oi.is_inverted() {
        return oi.lo <= i.hi || oi.hi >= i.lo;
    }
    oi.lo <= i.hi && oi.hi >= i.lo
}
